File: kotlin/src/dk/leetcode/BinaryTreePruning.kt
package dk.leetcode
// https://leetcode.com/problems/binary-tree-pruning/

class Solution_21 {
    fun pruneTree(root: TreeNode?): TreeNode? {
        if (root != null) {
            root.left = pruneTree(root.left)
            root.right = pruneTree(root.right)
            if (root.`val` != 1 && root.left == null && root.right == null)
                return null
        }
        return root
    }
}

File: c/lc/binary-tree-pruning.cc
// https://leetcode.com/problems/binary-tree-pruning/
class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        if (root) {
            root->left = pruneTree(root->left);
            root->right = pruneTree(root->right);
            if (root->left || root->right || root->val == 1) {
                return root;
            }
        }
        return NULL;
    }
};
